docarray/doc_index/backends/elastic_doc_index.py

```diff
@@ -0,0 +1,454 @@
+    import torch
+
+    ELASTIC_PY_VEC_TYPES.append(torch.Tensor)
+
+
+def _raise_has_substitue(name):
+    def _inner(self, *args, **kwargs):
+        raise NotImplementedError(
+            f'`{name}` is replaced by query() through the query builder of this Document Index ({type(self)}). '
+            f'You can also call `{type(self)}.{name}()` directly.'
+        )
+
+    return _inner
+
+
+class ElasticDocumentIndex(BaseDocumentIndex, Generic[TSchema]):
+    def __init__(self, db_config=None, **kwargs):
+        super().__init__(db_config=db_config, **kwargs)
+        self._db_config = cast(ElasticDocumentIndex.DBConfig, self._db_config)
+
+        if self._db_config.index_name is None:
+            id = uuid.uuid4().hex
+            self._db_config.index_name = 'index__' + id
+
+        self._index_name = self._db_config.index_name
+
+        self._client = Elasticsearch(
+            hosts=self._db_config.hosts,
+            **self._db_config.es_config,
+        )
+
+        # compatibility
+        self._server_version = self._client.info()['version']['number']
+        if int(self._server_version.split('.')[0]) >= 8:
+            os.environ['ELASTIC_CLIENT_APIVERSIONING'] = '1'
+
+        # ElasticSearh index setup
+        self._index_init_params = ('type',)
+        self._index_vector_params = ('dims',)
+
+        body: Dict[str, Any] = {
+            'mappings': {
+                'dynamic': True,
+                '_source': {'enabled': 'true'},
+                'properties': {},
+            }
+        }
+
+        for col_name, col in self._column_infos.items():
+            if not col.config:
+                continue  # do not create column index if no config is given
+            body['mappings']['properties'][col_name] = self._create_index(col)
+
+        if self._client.indices.exists(index=self._index_name):  # type: ignore
+            self._client.indices.put_mapping(
+                index=self._index_name, body=body['mappings']
+            )
+        else:
+            self._client.indices.create(index=self._index_name, body=body)
+
+        if len(self._db_config.index_settings):
+            self._client.indices.put_settings(
+                index=self._index_name, body=self._db_config.index_settings
+            )
+
+        self._refresh(self._index_name)
+
+    ###############################################
+    # Inner classes for query builder and configs #
+    ###############################################
+
+    class QueryBuilder(BaseDocumentIndex.QueryBuilder):
+        def __init__(self, outer_instance, **kwargs):
+            super().__init__()
+            self._search = Search(
+                using=outer_instance._client, index=outer_instance._index_name, **kwargs
+            )
+
+        def build(self, *args, **kwargs) -> Any:
+            return self
+
+        def query(self, *args, **kwargs):
+            self._search = self._search.query(*args, **kwargs)
+            return self
```

Comment:
Maybe it would be more in the spirit to create a new `QueryBuilder` with updated state at each operation? that way every step remains intact and is not modifed in-place. I think you @JoanFM argue for this pattern, right?